base/operators.jl

```diff
@@ -871,9 +871,40 @@ julia> fs = [
 julia> ∘(fs...)(3)
 3.0
 ```
+See also [ComposedFunction](@ref).
 """
 function ∘ end
 
+"""
+    ComposedFunction{F,G} <: Function
+
+Represents the composition of two callable objects `f::F` and `g::G`.
+```jldoctest
+julia> sin ∘ cos === Base.ComposedFunction(sin, cos)
+true
+
+julia> typeof(sin∘cos)
+Base.ComposedFunction{typeof(sin), typeof(cos)}
+
+julia> sin ∘ cos === Base.ComposedFunction(sin, cos)
+true
+```
+The composed pieces are stored in the fields of `ComposedFunction` and can be retrieved as follows:
+```jldoctest
+julia> composition = sin ∘ cos
+sin ∘ cos
+
+julia> composition.f === sin
```

Comment:
I would actually prefer to name the fields to e.g. `outer, inner` instead of `f,g`. It is hard to remember if this models `f∘g` or `g∘f`.

Reply:
I'm somewhat neutral about the choice of the field names. I see `f` and `g` as something like `x` and `y`. It's technically arbitrary but I've seen it many times so I don't think it's super hard to guess the ordering.

Just as a note, one (rather weak) upside of keep using `f` and `g` is that we can define `ComposedFunction` for older `julia` versions as done in DataFrames.jl (https://github.com/JuliaData/DataFrames.jl/pull/2274#discussion_r433460863). However, it doesn't matter much since 1.6 soon will be the LTS (and we can even define `getproperty` for the closure in Compat.jl).

Reply:
I think the order is not easy to guess at all. Just have a look at the [wikipedia article](https://en.wikipedia.org/wiki/Function_composition). There the order `g∘f` is more dominant, which is opposed to the current order we use.

Reply:
My personal first guess was btw `g∘f`.

Reply:
Well, I sometimes get inner/outer opposite somehow (though this is probably just me). Since you can always flip the arrows (and which direction feels natural depends on the context), I'm not sure if there are truly unambiguous names for them. But I'm likely nitpicking. I agree inner/outer is more descriptive for function composition.

Reply:
With `outer, inner` it is at least clear that `c::ComposedFunction` talks about `c.outer(c.inner(x))`. With `f,g` it is unclear whether we are talking about `c.f(c.g(x))` or `c.g(c.f(x))`. Remembering the direction of arrows is an independent problem that comes up in both scenarios.

Reply:
Btw, I also get inner outer wrong sometimes, but I can usually recover by pure reasoning without having to dig up the source.

Reply:
I agree it's easier to remember and be confident which function is applied first with inner/outer. I think I'm in the team inner/outer now. How about renaming them in this PR and see the reactions of other reviewers?

Reply:
For the little it is worth, I think `f` and `g` are fine and their exact interpretation can be clarified in the docstring. That way there doesn't need to be any work on Compat.jl or in other packages.

If you stick with inner and outer, then you need to change line 881:

```
Represents the composition of two callable objects `f::Outer` and `g::Inner`.
```

Reply:
No, `f` and `g` are confusing, we should not force people to look up the docstring all the time. Also @tkf has a PR to compat
https://github.com/JuliaLang/Compat.jl/pull/720 and supporting `f` and `g` is the easy part of that code.